Split core OTEL exporter from fastapi specific implementation

katybaulch opened the pull request:
# Description

Instead of having a single OTEL exporter which has generic & fastapi specific code needed for the otel exporters, split them into a generic and specialised class. This is key for the next PR where we introduce the OTEL prefect exporter, which we didn't want to instantiate fastapi specific stuff when it didn't need to.

This is just a refactor of the existing code Henry has written.

## Type of change

Please select the option(s) below that are most relevant:

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] GitHub workflow update
- [ ] Documentation update
- [x] Refactor legacy code
- [ ] Dependency update

## How Has This Been Tested?

Please describe the tests that you added to verify your changes.

## Reviewer Checklist

- [ ] **DB_CLIENT DEPENDENCY IS ON THE LATEST VERSION**
- [ ] The PR represents a single feature (small driveby fixes are also ok)
- [ ] The PR includes tests that are sufficient for the level of risk
- [ ] The code is sufficiently commented, particularly in hard-to-understand areas
- [ ] Any required documentation updates have been made
- [ ] Any TODOs added are captured in future tickets
- [ ] No FIXMEs remain

jamesgorrie commented on api/src/api/telemetry_base.py:
nit / question: can we keep class & filenames the same? i.e. `base_telemetry`? This question applies to all the files. It's a tiny, but annoying bit of cognitive load when importing things.

katybaulch replied:
done, everything is also exported in the init so you can just do `from api import BaseTelemetry, FastAPITelemetry etc`

jamesgorrie reviewed (commented):
I am battling to see what is happening here - but that's probably just because it is a large refactor and unfamiliarity with the code.

Are we:
- generating a BaseTelemetry class that we expect all OTel instrumentation to extend / inherit from?
- the first instance we have is `FastAPITelemetry`?

If this is right - could we use this as a chance to describe what the abstraction is actually doing, and describe how it's useful?

100% sure it is, but it is not massively evident to me as to how - but I'm assuming that is a knowledge gap that maybe some docs or comments might help with

katybaulch commented:
> I am battling to see what is happening here - but that's probably just because it is a large refactor and unfamiliarity with the code.
> 
> Are we:
> 
> * generating a BaseTelemetry class that we expect all OTel instrumentation to extend / inherit from?
> * the first instance we have is `FastAPITelemetry`?
> 
> If this is right - could we use this as a chance to describe what the abstraction is actually doing, and describe how it's useful?
> 
> 100% sure it is, but it is not massively evident to me as to how - but I'm assuming that is a knowledge gap that maybe some docs or comments might help with

You got it! We have extracted the shared OpenTelemetry bootstrap that Henry wrote into a BaseTelemetry class, which handles the boring but essential and generic bits of the OTel setup: tracer/logger initialisation, OTLP wiring & exporting, and span enrichment. Framework layers, of which FastAPI is the first, now subclass BaseTelemetry to bolt on their specific concerns (async exception hooks, route wrapping, prefect logger hijacking etc.) whilst inheriting the consistent base config. I have added module level docs to explain that shape so the abstraction is clearer
